XenAdmin/Controls/SrPicker.cs

```diff
@@ -118,111 +118,198 @@ protected override void OnPaint(PaintEventArgs e)
-                Program.Invoke(this, () => Rebuild());
+            if (sender is SR sr &&
+                (e.PropertyName == "name_label" || e.PropertyName == "PBDs" ||
+                 e.PropertyName == "physical_utilisation" || e.PropertyName == "virtual_allocation"))
+            {
+                Program.Invoke(this, () =>
+                {
+                    foreach (var item in Items)
+                    {
+                        if (item is SrPickerItem it && !it.Scanning && it.TheSR.opaque_ref == sr.opaque_ref)
+                        {
+                            it.Update();
+                            break;
+                        }
+                    }
+                });
+            }
         }
 
         private void SR_CollectionChanged(object sender, CollectionChangeEventArgs e)
         {
-            if (_scanCount > 0)
+            if (e.Action == CollectionChangeAction.Add && e.Element is SR addedSr)
+            {
+                Program.Invoke(this, () =>
+                {
+                    foreach (var item in Items)
+                    {
+                        if (item is SrPickerItem it && it.TheSR.opaque_ref == addedSr.opaque_ref)
+                            return;
+                    }
+
+                    AddNewSr(addedSr);
+                });
                 return;
+            }
+            
+            if (e.Action == CollectionChangeAction.Remove)
+            {
+                var removedSrs = new List<string>();
+
+                if (e.Element is SR storage)
+                    removedSrs.Add(storage.opaque_ref);
+                else if (e.Element is List<SR> range)
+                    removedSrs = range.Select(sr => sr.opaque_ref).ToList();
+                else
+                    return;
+
+                Program.Invoke(this, () =>
+                {
+                    var itemsToRemove = new List<SrPickerItem>();
+
+                    foreach (var item in Items)
+                    {
+                        if (item is SrPickerItem it && removedSrs.Contains(it.TheSR.opaque_ref))
+                        {
+                            foreach (var pbdRef in it.TheSR.PBDs)
+                            {
+                                var pbd = it.TheSR.Connection.Resolve(pbdRef);
+                                if (pbd != null)
+                                    pbd.PropertyChanged -= pbd_PropertyChanged;
+                            }
+                            it.TheSR.PropertyChanged -= sr_PropertyChanged;
+
+                            it.ItemUpdated -= Item_ItemUpdated;
+                            itemsToRemove.Add(it);
+                            break;
+                        }
+                    }
 
-            Program.Invoke(this, () => Rebuild());
+                    foreach (var item in itemsToRemove)
+                        RemoveNode(item);
+                });
+            }
         }

```

Comment:
`SR_CollectionChanged` is called by other threads indeed, but because we call Invoke in it, the places where we enumerate the list items and Add/RemoveNode are only accessed by the main thread.
The lock inside AddNewSr is protecting the refreshQueue. At that particular line the queue is accessed by the main thread, but elsewhere it may be accessed by other threads, like at the line where actions finishing are removed, hence the lock.

Reply:
Cool make sense. Thanks for the clarification. LGTM